doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,37 @@
+# Design Doc: Execute the Program with Multi Thread
+
+## Abstract
+
+This Design Doc propose an approach to make the user-defined Op graph
+running with multi threads, we will use an auto transpiler to convert the user-defined
+Op graph to a multi-thread Op graph, and run `ParallelDo` Op to run the
+multi-thread graph.
+
+## Graph Converter
+<img src="src/multi-threads/single-thread@3x.png" width="300">
+
+After converted:
+
+<img src="src/multi-threads/multi-threads@3x.png" width="1000">
+
+## Implement
+
+- `Multi-Thread Transpiler`  will convert the graph to a multi-threads graph
+    which would be executed with multi-threads.
+- `BlockingCounter` will `Init/Decrement` a condition variable, and Blocking `Wait`
```

Comment:
Name to `ConditionVariable` should be fine. `ConditionVariable` can store any type of "condition", not only counter.

Reply:
I think the `ConditionVariable` is a great idea, how about adding this type in variable in fluid?  And it also needs some Op like `condition_varaible.wait()/notify_all()/notify_one()`. How about adding a design doc for `Condition Variable` firstly?